src/librustc/diagnostics.rs

```diff
@@ -303,10 +303,63 @@ number cannot be negative.
 E0307: r##"
 The length of an array is part of its type. For this reason, this length must be
 a compile-time constant.
+"##,
+
+E0308: r##"
+This error occurs when the compiler was unable to infer the concrete type of a
+variable. This error can occur for several cases, the most common of which is
+that there is a mismatch in the expected type that the compiler inferred, and
+the actual type that the user defined a variable as.
```

Comment:
I agree with @cactorium  that original is grammatically correct but also difficult to read.

I suggest perhaps:
> ... the most common of which is a mismatch in the expected type that the compiler inferred for a variable's initializing expression, and the actual type explicitly assigned to the variable.
> 
> For example:
> ```rust
> let x: i32 = "I am not a number!";
> //     ~~~   ~~~~~~~~~~~~~~~~~~~~
> //      |             |
> //      |    initializing expression;
> //      |    compiler infers type `&str`
> //      |
> //    type `i32` assigned to variable `x`
> ```


(Or drop the example if you're not into that. :)  )